handler/src/main/java/io/netty/handler/ssl/PemX509Certificate.java

```diff
@@ -0,0 +1,382 @@
+ * to do crypto stuff in Java land.
+ *
+ * All methods other than what's implemented in {@link PemEncoded}'s throw
+ * {@link UnsupportedOperationException}s.
+ *
+ * @see PemEncoded
+ * @see OpenSslContext
+ */
+public final class PemX509Certificate extends X509Certificate implements PemEncoded {
+
+    private static final byte[] BEGIN_CERT = "-----BEGIN CERTIFICATE-----\n".getBytes(CharsetUtil.US_ASCII);
+    private static final byte[] END_CERT = "\n-----END CERTIFICATE-----\n".getBytes(CharsetUtil.US_ASCII);
+
+    /**
+     * Creates a {@link PemEncoded} value from the {@link X509Certificate}s.
+     */
+    static PemEncoded toPEM(ByteBufAllocator allocator, boolean useDirect,
+            X509Certificate[] chain) throws CertificateEncodingException {
+
+        // We can take a shortcut if there is only one certificate and
+        // it already happens to be a PemX509Certificate instance.
+        if (chain.length == 1) {
+            X509Certificate first = chain[0];
+            if (first instanceof PemEncoded) {
+                return ((PemEncoded) first).retain();
+            }
+        }
+
+        boolean success = false;
+        ByteBuf pem = null;
+        try {
+            for (X509Certificate cert : chain) {
+                if (cert instanceof PemEncoded) {
+                    ByteBuf content = ((PemEncoded) cert).content();
+
+                    if (pem == null) {
+                        pem = newBuffer(allocator, useDirect, content.readableBytes() * chain.length);
+                    }
+
+                    pem.writeBytes(content.slice());
+                } else {
+                    ByteBuf encoded = Unpooled.wrappedBuffer(cert.getEncoded());
+                    try {
+                        ByteBuf base64 = SslUtils.toBase64(allocator, encoded);
+                        try {
+                            if (pem == null) {
+                                // Approximate the total size. The sizes for certificates can vary a lot
+                                // but the "normal" use-case is that there's only one certificate in which
+                                // case we'll be spot on!
+                                pem = newBuffer(allocator, useDirect,
+                                        (BEGIN_CERT.length + base64.readableBytes() + END_CERT.length) * chain.length);
+                            }
+
+                            pem.writeBytes(BEGIN_CERT);
+                            pem.writeBytes(base64);
+                            pem.writeBytes(END_CERT);
+                        } finally {
+                            base64.release();
+                        }
+                    } finally {
+                        encoded.release();
+                    }
+                }
+            }
+
+            if (pem == null) {
+                throw new IllegalStateException();
+            }
+
+            success = true;
+            return new PemValue(pem, false);
```

Comment:
if the constructor of PemValue were to throw for some reason, it seems like we would leak the buffer. Consider adding comments on the constructor that it must take ownership of the buffer, or handle the release here:

```java
PemEncoded result = new PemValue(pem, false);
success = true;
return result;
```